lib/core/src/Cardano/Wallet/Primitive/Types/UTxOIndex/Internal.hs

```diff
@@ -245,20 +258,24 @@ insertMany = flip $ F.foldl' $ \u (i, o) -> insert i o u
 -- of applying this function will be equivalent to the identity function.
 --
 delete :: TxIn -> UTxOIndex -> UTxOIndex
-delete i u = case Map.lookup i (utxo u) of
-    Nothing -> u
-    Just o -> UTxOIndex
-        { index = F.foldl' deleteEntry (index u) (txOutAssets o)
-          -- This operation is safe, since we have already determined that the
-          -- entry is a member of the index, and therefore the balance must be
-          -- greater than or equal to the value of this output:
-        , coins = Set.delete i (coins u)
-        , balance = balance u `TokenBundle.unsafeSubtract` view #tokens o
-        , utxo = Map.delete i $ utxo u
-        }
+delete i u =
+    maybe u updateIndex $ Map.lookup i $ utxo u
   where
-    txOutAssets :: TxOut -> Set AssetId
-    txOutAssets = TokenBundle.getAssets . view #tokens
+    updateIndex :: TxOut -> UTxOIndex
+    updateIndex o = u
+        -- This operation is safe, since we have already determined that the
+        -- entry is a member of the index, and therefore the balance must be
+        -- greater than or equal to the value of this output:
+        & over #balance (`TokenBundle.unsafeSubtract` view #tokens o)
+        & over #utxo (Map.delete i)
+        & case categorizeTxOut o of
+            IsCoin ->
+                over #coins (Set.delete i)
+            IsCoinWithSingletonAsset a -> id
+                . over #assetsSingleton (`deleteEntry` a)
+                . over #assetsAll (`deleteEntry` a)
```

Comment:
Ok. So, `assetAll` also contain singleton assets (as stated in a comment above :+1:).

Reply:
> Ok. So, assetAll also contain singleton assets (as stated in a comment above +1).

I did have a debate with myself about making two disjoint sets: `assetsMixed` and `assetsSingleton`. But that might make it harder for callers that don't care if an asset is accompanied by other assets, since they'd have to specify two selection filters to get what they wanted. In the end, it seemed nicer to treat `assetsSingleton` as a subset of `assetsAll`, similar to how `WithAdaOnly` is a subset of `Any`.